### DISCUSSION TITLE
how contributors are supposed to debug their changes?

### DISCUSSION THREAD

**Discussion by tbolon** [dotnet/winforms#4369] [Q&A] [answered]
I created an [issue](https://github.com/dotnet/winforms/issues/4366) regarding the documentation, but perhaps a discussion is more suited for that.

I am asking feedback about how external contributors are supposed to test changes in the repository ?

I see there is also a WinformsControlsTest project, but I could not open any Form using the VS designer because every time I double click on a form to open the designer, I receive the error "Timed out while connecting to named pipe.".

Do you have any tips or suggestion about how to test changes in the core controls ?

**Comment by weltkante** [comment, marked as answer]
I don't think there is any design time support for custom WinForms builds, personally I just add some code for testing to the WinFormsControlTest project you already found (usually by replacing the main form invocation) and run the project. Make sure you start VS via `start-vs.cmd` to make verything work.

**Reply by tbolon** [reply]
I have done exactly the same, manually adding code in the form ctor to test my changes. Thanks.

**Comment by RussKie** [comment]
> but I could not open any Form using the VS designer because every time I double click on a form to open the designer, I receive the error "Timed out while connecting to named pipe.".

Yes, unfortunately the designer is unable to load the canary build. Not unless you copy those into the global dotnet folders (under program files).
The team is aware of this, but so far there hasn't been much traction.

Never the less the changes can be tested at runtime. 
The design portion is cumbersome, but you can either design a form manually, or in the "classic" designer in a project targeting .NET Framework, or have a multi-targeting project. These are obviously workarounds.

And as @weltkante said, we have `start-code.cmd` and `start-vs.cmd` helper scripts that set env variables so that VS knows where the target runtime is.
